src/main/lib/osu-file-parser/OsuParser.ts

```diff
@@ -1,125 +1,345 @@
+};
+
+Buffer.prototype.read_u32 = function () {
+  const bytes = this.read_bytes(4);
+  return bytes.readUInt32LE(0);
+};
+
+Buffer.prototype.read_u16 = function () {
+  const bytes = this.read_bytes(2);
+  return bytes.readUInt16LE(0);
+};
+
+Buffer.prototype.read_u8 = function () {
+  return this.read_bytes(1)[0] & 0xff;
+};
 
 export class OsuParser {
-  static async parseDir(dir: string, update?: (i: number, total: number, file: string) => any): DirParseResult {
-    if (!await access(dir, fs.constants.R_OK)) {
-      return fail('Directory does not exists.');
+  static async parseDb(
+    dbpath: string,
+    update?: (i: number, total: number, file: string) => any
+  ): DirParseResult {
+    let db;
+    let songsFolderPath = dbpath + "/Songs";
+
+    try {
+      // NOTE: This isn't readFile from fs-promises.ts.
+      //       We want to read binary data here, not utf-8 encoded data!
+      db = await fs.promises.readFile(dbpath + "/osu!.db");
+    } catch (err) {
+      return fail("Failed to read osu!.db.");
     }
 
-    let lastAudioID = "";
-    let lastAudioSong: any = undefined;
+    // Scan for cfg file to check for a custom songs folder path
+    try {
+      const username = os.userInfo().username;
+      const cfgFile = await fs.promises.readFile(`${dbpath}/osu!.${username}.cfg`, "utf-8");
+      const lines = cfgFile.split("\n");
+      console.log("cfg file found");
+
+      for (const line of lines) {
+        if (line.includes("BeatmapDirectory")) {
+          const customPath = line.split("=")[1].trim();
+          if (customPath !== "Songs") {
+            songsFolderPath = customPath;
+            console.log(`Custom songs folder path found: ${songsFolderPath}`);
+          }
+        }
+      }
+    } catch (err) {
+      console.error(`Either no cfg file was found or there was an error reading it: ${err}`);
+    }
 
-    // const audioSources = new Set<string>();
-    const dirs = await getSubDirs(dir);
     const songTable = new Map<ResourceID, Song>();
     const audioTable = new Map<ResourceID, AudioSource>();
-    const imageTable = new Map<ResourceID, ImageSource>();
-
-    for (let i = 0; i < dirs.length; i++) {
-      const subDirPath = path.join(dir, dirs[i]);
 
-      const files = await getFiles(subDirPath, "osu");
-      for (let j = 0; j < files.length; j++) {
-        if (update !== undefined) {
-          update(i + 1, dirs.length, files[j]);
-        }
+    // NOTE: Images can only be fetched from the .osu file AFAIK.
+    //       You should parse those dynamically when needed.
+    const imageTable = new Map<ResourceID, ImageSource>();
 
-        const file = path.join(subDirPath, files[j]);
-        const audioSrc = await OsuFile.getProp(file, "AudioFilename");
+    // HACK: Used by prototype overrides above
+    db.pos = 0;
 
-        if (audioSrc.isError) {
-          continue;
-        }
+    const db_version = db.read_u32();
+    console.debug("db version:", db_version);
```

Comment:
All console.debug should be changed to proper logging or removed

Reply:
went ahead and removed the debugs